src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1389 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
+{
+    [CompilerTrait(CompilerFeature.AsyncStreams)]
+    public class CodeGenAsyncDisposableTests : CSharpTestBase
+    {
+        private static readonly string s_interfaces = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.Task DisposeAsync();
+    }
+}
+";
+
+        [Fact]
+        public void TestWithCSharp7_1()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    async System.Threading.Tasks.Task M()
+    {
+        using await (var x = new C())
+        {
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, parseOptions: TestOptions.Regular7_1);
+            comp.VerifyDiagnostics(
+                // (6,15): error CS8302: Feature 'async streams' is not available in C# 7.1. Please use language version 7.2 or greater.
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_1, "await").WithArguments("async streams", "7.2").WithLocation(6, 15)
+                );
+            // PROTOTYPE(async-streams) LangVersion for async-streams will be adjusted before release
+        }
+
+        [Fact]
+        public void TestInNonAsyncVoidMethod()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    void M()
+    {
+        using await (var x = new C())
+        {
+            return;
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (6,22): error CS4033: The 'await' operator can only be used within an async method. Consider marking this method with the 'async' modifier and changing its return type to 'Task'.
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutVoidAsyncMethod, "var x = new C()").WithLocation(6, 22)
```

Comment:
>var x = new C() [](start = 74, length = 15)

I think it would be better if the location for the error was the ```await``` keyword. It would point precisely to the problem and the part between parenthesis could be quite long, event could span multiple lines.